### PR TITLE
DHFPROD-5974: Ingest documents with user-defined prefix and guaranteed unique URI

### PR DESCRIPTION
**Pull request by anu3990**
### Description
This pull request focuses on the changes needed for - https://project.marklogic.com/jira/browse/DHFPROD-5974

#### Checklist: 
```diff
- Note: do not change the below
```

-  ##### Owner:

- [x] JIRA_ID included in all the commit messages
- [x] PR title is in the format JIRA_ID:Title
- [x] Rebase the branch with upstream
- [x] Squashed all commits into a single commit
- [x] Added Tests
  

- ##### Reviewer:

- [x] Reviewed Tests

### CODE REVIEW

**Review by SameeraPriyathamTadikonda — APPROVED** [review]

**Review by rjrudin — DISMISSED** [review]

**Comment by anu3990** [comment]
Rebased with the develop branch and also fixed an additional test that was failing - "BulkIngestTest.java"

File: marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataTest.java [hunk 1]
```diff
@@ -91,10 +94,24 @@ private void verifyFruitCount(int expectedCount, String message) {
...
-            assertEquals("apple", doc.get("envelope").get("instance").get("fruitName").asText());
-            assertEquals("red", doc.get("envelope").get("instance").get("fruitColor").asText());
+        String uriQuery = "cts.uriMatch('/testFruit/**')";
+        EvalResultIterator uriQueryResult = getHubClient().getStagingClient().newServerEval().javascript(uriQuery).eval();
+        AtomicBoolean foundRedApple = new AtomicBoolean(false);
+        AtomicBoolean foundYellowBanana = new AtomicBoolean(false);
+        uriQueryResult.iterator().forEachRemaining(item -> {
+            final String expectUri = item.getString();
+            JsonNode doc = getHubClient().getStagingClient().newJSONDocumentManager().read(expectUri, new JacksonHandle()).get();
+            if((doc.get("envelope").get("instance").get("fruitName").asText().equals("apple")) &&
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataTest.java:104]: This test is a little brittle in that if the structure of the document doesn't match, we'll get a null-pointer exception. 

Instead of doing a cts.uriMatch, I recommend doing a cts.uris query like this:

    cts.uris(null, [limit='1'], cts.andQuery([
      cts.directoryQuery("/testFruit/"),
      cts.jsonPropertyValueQuery("fruitName", someVariable)
    ]))

You can use newServerEval to execute that and then verify that you get 1 result back and that it starts with "/testFruit/". 

That results in a much simpler test and also much less chance of a null-pointer occurring.

**Review by rjrudin — CHANGES_REQUESTED** [review]

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/ingestion/bulkIngester.sjs [hunk 2]
```diff
@@ -36,7 +36,7 @@ const inputs =
       [{UNKNOWN: input}];
 inputs.forEach(record => {
   state.next = state.next + 1;
-  const uri = (state.prefix) + '/' + (work.taskId) + '/' + (state.next) + '.json';
+  const uri = (state.uriprefix) + '/' + sem.uuidString() + '.json';
```
Comment (rjrudin) [inline marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/data-services/ingestion/bulkIngester.sjs:39]: Just realized that we should leave out "/". That allows Ernie to have URIs of the form "(uuid).json" if he wants. That's consistent with how MLCP works as well - i.e. no "/" is added unless the user configures one.

**Review by rjrudin — CHANGES_REQUESTED** [review]
Looks good, just identified one small tweak.

**Review by rjrudin — APPROVED** [review]

**Comment by SameeraPriyathamTadikonda** [comment]
I think we should write another test to verify how connector behaves when the user does not provide "uriprefix"

**Review by rjrudin — CHANGES_REQUESTED** [review]
That's a good point by @SameeraPriyathamTadikonda - we want another test scenario (separate test method) where uriprefix is not set at all, and then we verify that the doc is still written and the URI ends with ".json" and does not start with "/".

**Review by SameeraPriyathamTadikonda — APPROVED** [review]

**Review by SameeraPriyathamTadikonda — DISMISSED** [review]

File: marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataTest.java [hunk 3]
```diff
@@ -48,6 +47,25 @@ void ingestThreeFruitsWithBatchSizeOfTwo() throws IOException {
...
+    @Test
+    public void testBulkIngestWithoutUriPrefix() throws IOException {
+        DataWriter<InternalRow> dataWriter = buildDataWriter("1", "");
+        dataWriter.write(buildRow("pineapple", "green"));
+
+        String uriQuery = "cts.uris('', null, cts.andQuery([\n" +
+            "  cts.jsonPropertyValueQuery('fruitName', 'pineapple')\n" +
+            "]))";
+
+        EvalResultIterator uriQueryResult = getHubClient().getStagingClient().newServerEval().javascript(uriQuery).eval();
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataTest.java:60]: No need to change here, but most of the time, it's sufficient to just do this:

    String uri = getHubClient().getStagingClient().newServerEval().javascript(query).evalAs(String.class);

If that fails because there's not a URI, that's totally fine. And we don't need to worry about verifying that there's not another URI - we know the database was cleaned before the test ran, so it's safe to assume there's one or zero URIs.

File: marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataTest.java [hunk 4]
```diff
@@ -48,6 +47,25 @@ void ingestThreeFruitsWithBatchSizeOfTwo() throws IOException {
...
+        assertTrue(uriQueryResult.hasNext());
+        String uri = uriQueryResult.next().getString();
+
+        assertTrue(uri.endsWith(".json"));
+        assertFalse(uri.startsWith("/"));
```
Comment (rjrudin) [inline marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/WriteDataTest.java:65]: Assertion calls are often the best place to document *why* something should be the way it is. In this case, there's a question of - well why doesn't it start with "/"? So I'd do:

    assertFalse(uri.startsWith("/"), "If the user wants the URI to start with a forward slash, the user must provide one. If the user doesn't, then it's assumed that the user doesn't want a forward slash at the start of the URI, so the endpoint will not add one automatically.");

That makes it very clear why it's expected. This isn't needed for every assertion, but this is a good one to document in the above manner.

**Review by rjrudin — CHANGES_REQUESTED** [review]

**Review by rjrudin — APPROVED** [review]

**Review by SameeraPriyathamTadikonda — APPROVED** [review]
